### DISCUSSION TITLE
Ideas for snow patch detection

### DISCUSSION THREAD

**Discussion by ipoole** [SimonFisher92/Scottish_Snow_Satellite_Data#24] [General]
This follows on from a discussion between Simon, Murray and Ian yesterday (3/12).   Thought it more convenient and open to put down thoughts here rather than in an email chain.

## Ground-truth collection
Simon believes there are people interested in labelling images for snow patches.   A couple of questions arise immediately:

1. What infra-structure?   Rolling our own could be a lot of work, especially if we want it to be robust in the hands of non-experts.  There is not only the graphical tools, but managing the inputs, results, authentication, allocation....  Collecting pixel level labels in a crowd sourced fashion is such a common and largely generic requirement, there must surely be a web based tool / service.   How about Amazon SageMaker for example?  I haven't yet looked at this in any detail, there may be costs.
2. What information do the GT labellers have?    RGB images presumably.   How about contour maps (implying height and slope orientation)?  Week of the year?   

## Training a classifier
Similar to the above, what inputs should the ML be given?   In terms of Sentinel bands, presumably:

- RGB - bands 2,3,4
- Near-infrared - band 8 (and 8A?)
- Water vapor - band 9
- Cirrus - band 10.

What about altitude, which is a CNN context provides slope?   Week of the year?   (It is amusing to think how a classifier would perform given this information only!).

Do we also give the classifier the Copernicus 'snow mask'?   I think there can be no harm.  It would be better if we had a greyscale version of this (probability image), pre-threshould.

Murray has suggested starting with a simple per-pixel classifier, which I think makes sense.   A simple quadratic (MVG) or decision forest.    

If we move on to a CNN, do we start from scratch or use a pre-trained model?   But the latter approach is only possible for a 3-band RGB input, is that right?   Training a CNN from scratch with so few images (~100 ?) and very few +ve pixels may be problematic.

**Comment by ipoole** [comment]
I'm looking some more at SageMaker.   I'm pretty sure it does what we need.   I'm trying to setup a 'job', for image segmentation.  I've uploaded some example .png files to an S3 bucket, but I'm getting tangled up with AWS permissions, IAM roles etc....   Would anyone like to pair with me to get this sorted?   I'm mostly available all this week.

**Comment by ipoole** [comment]
I'm slowly figuring out SageMaker.   Like all things AWS there are many levels of indirection around users, roles, policies, permissions....   Anyway I have successfully setup a SageMaker domain with access to some example snow patches on S3.   A gotcha which held me up for hours is that the AWS region matters....  Long story short ... although the "Sockholm" region does support SageMaker, the "Ground Truth" facility is mysteriously and silently absent!   So I switched to "US East (N. Virginia)" where suddenly it appears (???).  You also have to ensure the S3 bucked is in the same region...    So many wheels...

Anyway, I'm convinced that the SageMaker Ground Truth labelling will do what we want.   I need another session (maybe later today) to complete the labelling job setup and workers list...   Then hope to share some images.  Of course, the main purpose of SageMaker is to conveniently train models, so we might try that too eventually.

**Comment by ipoole** [comment]
I have a basic per-pixel classifier going, trained on the SCL data.    Evaluation is by robustly assigning whole patch areas to train or test.  A quadratic discriminant (aka multivariate Gaussian) gives AUC = 0.94.   A decision forest gives 0.95 but takes several minutes to train, hardly worth it.    Here are some image results:

![image](https://github.com/SimonFisher92/Scottish_Snow/assets/10311507/270b273c-4205-4285-b215-abf846c6f3ad)

The SCL GT has used some morphology to yield more spatially coherent results.  We could do the same ... if we can be bothered.
After committing this (to branch 27) I'll put it down and return to the SageMaker GT tool.

**Reply by murraycutforth** [reply]
This is awesome @ipoole, great stuff! So is this classifier just using the 13 bands as input? No elevation or slope data? I'm interested in your overall opinion of the SCL annotations now that you've looked at them in depth. I was very skeptical of their quality earlier on in this project as I could see many cases of the SCL being obviously wrong, although some of this was other classes such as water, which we don't care about. EDIT- sorry for repeating Simon's question about the model inputs..

**Reply by murraycutforth** [reply]
Anyway if the SCL masks do look to be accurate enough for us, that's great news, as we can get on with using them to start answering some "research" questions.

**Comment by SimonFisher92** [comment]
Thanks for sharing this Ian. I find this to be very convincing. My main concern was discrimination between cloud and snow but it's doing a very good job not making that mistake. Did you manage to get the topology maps in there? As in, the colour maps I generated? To be honest it doesn't even really look like we need them. For some pixels it actually looks like GT is wrong and the pred is correct. 

Are you able to get coefficients out? Id be interested to see what the primary predictive features are.

Mind if I share this with that PhD student whos PhD is on Scottish snow patches and satellite data?

**Reply by ipoole** [reply]
Hi Simon, thanks for your comments.   

Indeed clouds are a big issue.   I selected patches (for both training and test) which had <20% cloud.   Is that right?   I also tried introducing cloud as another class (so a three class problem), which made little difference.

No I haven't used the topology information.   I can believe this would help when evaluated against gold standard GT, but I'm not sure there is much point when evaluating against SLC, which didn't have this information.    Still, it probably should be included.   How closely can we match SLC with *only* topology information?!
 
We must remember what we're doing here, which is simply to reverse engineer the SCL algorithm by ML.   We cannot do better that SCL by this approach, because that is what it's trained on.   The pixel classifier has ~5% disagreement.   It's wishful thinking to hope that in  those 5% we are "right" and SCL is "wrong" (the quotes implying appeal to the great ground-truther in the sky which embodies absolute gold standard truth).     

So there are many questions about how we should proceed.  Do we invest in creating our own GT.  ... probably needs another discussion.

**Comment by ipoole** [comment]
**GageMaker GT labelling**

After much frustration I have created a SageMaker labelling job.   You (Simon, Murray) should have received an invitation to join a snow patches labelling job.   You aught to be able to login using the link and try the labelling interface.  There are just a handful of example images at the moment.
I've labelled three images myself, and can already see this is not easy and there are several issues we need to consider.   Should we explicitly label cloud?   How spatially accurate do we need to be?   How comprehensive?   Do we include mostly cloudy images in the image set?

I feel it will be a substantial task to obtain GT which surpass the SCL labels in quality, and approach it in quantity.

**Reply by murraycutforth** [reply]
Another idea related to this which I wanted to pitch, is an annotation effort where instead of annotating snow ourselves we just look at the existing SCL masks and manually indicate whether it's a good snow/cloud segmentation or not. Since it's just image-level annotation it's relatively low effort. Then we can use this higher-quality subset of the SCL masks to train our own model.

**Reply by ipoole** [reply]
Hi @murraycutforth.  Yes that SLC mask correcting approach is an option.   I'm not sure if SageMaker supports that, need to check.   Such approaches do tend to introduce a sort of algorithmic bias .... I've spoken of this before I'm sure, with the example of correcting automatically generated vessel centerlines.   Because many (some) cases (in our case pixels) will be unclear, they will tend to be left unchanged.  So the original algorithm will tend to be favoured.    That said, it may still make sense - if SageMaker supports it, otherwise there might be substantial work developing our own process.
